### PR TITLE
HDDS-5903. Add Support for Bucket Owner Acls

### PR DESCRIPTION
**Pull request by aswinshakil**
## What changes were proposed in this pull request?

To add `owner` field to Buckets and support `checkAcls` for the new Bucket Owner.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-5903

## How was this patch tested?

This patch was tested manually, integration tests, and CI tests.
https://github.com/aswinshakil/ozone/actions/runs/1442975996

### CODE REVIEW

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java [hunk 1]
```diff
@@ -2083,6 +2101,44 @@ private String getVolumeOwner(String volume) throws OMException {
     }
   }
 
+  public String getBucketOwner(String volume, String bucket, ACLType type,
```
Comment (smengcl) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:2104]: Add javadoc to this method

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java [hunk 2]
```diff
@@ -39,10 +40,7 @@
 import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos.LayoutVersion;
 import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos.OMRequest;
 import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos.OMResponse;
-import org.apache.hadoop.ozone.security.acl.IAccessAuthorizer;
-import org.apache.hadoop.ozone.security.acl.OzoneObj;
-import org.apache.hadoop.ozone.security.acl.OzoneObjInfo;
-import org.apache.hadoop.ozone.security.acl.RequestContext;
+import org.apache.hadoop.ozone.security.acl.*;
```
Comment (smengcl) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java:43]: Try not to allow IntelliJ to automatically use wildcard imports:

https://www.jetbrains.com/help/idea/creating-and-optimizing-imports.html#disable-wildcards-for-class

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/bucket/CreateBucketHandler.java [hunk 3]
```diff
@@ -51,6 +52,10 @@
           "false/unspecified indicates otherwise")
   private Boolean isGdprEnforced;
 
+  @Option(names = {"--user", "-u"},
+          description = "Owner of the bucket")
```
Comment (smengcl) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/bucket/CreateBucketHandler.java:56]:
```suggestion
          description = "Owner of the bucket. Defaults to current user if not specified")
```

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java [hunk 4]
```diff
@@ -2083,6 +2101,44 @@ private String getVolumeOwner(String volume) throws OMException {
...
+    String dbBucketKey = metadataManager.getBucketKey(volume, bucket);
+    OmBucketInfo bucketInfo = null;
+    try {
+      bucketInfo = metadataManager.getBucketTable().get(dbBucketKey);
+    } catch (IOException ioe) {
+      if (ioe instanceof OMException) {
+        throw (OMException)ioe;
+      } else {
+        throw new OMException("getBucketOwner for Bucket " + volume + "/" +
+                bucket  + " failed", ResultCodes.INTERNAL_ERROR);
```
Comment (smengcl) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:2128]: Better include `ioe.getMessage()` here as well in the exception message to make potential troubleshooting easier.
```suggestion
        throw new OMException("getBucketOwner for Bucket " + volume + "/" +
                bucket  + " failed: " + ioe.getMessage(), ResultCodes.INTERNAL_ERROR);
```

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java [hunk 5]
```diff
@@ -234,15 +233,66 @@ protected void checkACLs(OzoneManager ozoneManager, String volumeName,
...
+      UserGroupInformation currentUser = createUGI();
+      contextBuilder.setClientUgi(currentUser);
       contextBuilder.setIp(getRemoteAddress());
       contextBuilder.setHost(getHostName());
       contextBuilder.setAclType(IAccessAuthorizer.ACLIdentityType.USER);
-      contextBuilder.setOwnerName(volumeOwner);
+
+      boolean isVolOwner = isOwner(currentUser, volumeOwner);
+      IAccessAuthorizer.ACLType parentAclRight = aclType;
+      if(isVolOwner) {
```
Comment (smengcl) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java:247]:
```suggestion
      if (isVolOwner) {
```

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java [hunk 6]
```diff
@@ -234,15 +233,66 @@ protected void checkACLs(OzoneManager ozoneManager, String volumeName,
...
+        contextBuilder.setOwnerName(volumeOwner);
+      } else {
+        contextBuilder.setOwnerName(bucketOwner);
+      }
+      if(ozoneManager.isNativeAuthorizerEnabled()) {
```
Comment (smengcl) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java:252]:
```suggestion
      if (ozoneManager.isNativeAuthorizerEnabled()) {
```

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java [hunk 7]
```diff
@@ -234,15 +233,66 @@ protected void checkACLs(OzoneManager ozoneManager, String volumeName,
...
+        if (aclType == IAccessAuthorizer.ACLType.CREATE ||
+                aclType == IAccessAuthorizer.ACLType.DELETE ||
+                aclType == IAccessAuthorizer.ACLType.WRITE_ACL) {
+          parentAclRight = IAccessAuthorizer.ACLType.WRITE;
+        } else if (aclType == IAccessAuthorizer.ACLType.READ_ACL ||
+                aclType == IAccessAuthorizer.ACLType.LIST) {
+          parentAclRight = IAccessAuthorizer.ACLType.READ;
+        }
+      } else {
+        parentAclRight =  IAccessAuthorizer.ACLType.READ;
```
Comment (smengcl) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java:262]:
```suggestion
        parentAclRight = IAccessAuthorizer.ACLType.READ;
```

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java [hunk 8]
```diff
@@ -234,15 +233,66 @@ protected void checkACLs(OzoneManager ozoneManager, String volumeName,
...
     }
   }
 
+  private boolean isOwner(UserGroupInformation callerUgi,
+                                 String ownerName) {
+    if (ownerName == null) {
+      return false;
+    }
+    if (callerUgi.getUserName().equals(ownerName) ||
+            callerUgi.getShortUserName().equals(ownerName)) {
```
Comment (smengcl) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java:290]:
```suggestion
        callerUgi.getShortUserName().equals(ownerName)) {
```

**Review by smengcl — COMMENTED** [review]
Good work @aswinshakil . Overall lgtm. I have 3 comments / questions, apart from the inline ones:

1. Pls resolve the merge conflict in `RpcClient.java`.
2. Do we need to check and replace all the other `checkAcls()` calls to `OzoneAclUtils.checkAllAcls()` ?
3. Are all the comments in the previous PR #2771 addressed ?

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestBucketOwner.java [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+    // Test Key Operations Non-Bucket Owner, Non-Volume Owner
+    //Key Create
+    UserGroupInformation.setLoginUser(user3);
+    OzoneBucket ozoneBucket;
+    try {
+      OzoneVolume volume = cluster.getClient().getObjectStore()
+              .getVolume("volume1");
+      ozoneBucket = volume.getBucket("bucket1");
+      createKey(ozoneBucket, "key3", 10, new byte[10]);
+      fail();
```
Comment (smengcl) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestBucketOwner.java:129]:
```suggestion
      fail("Create key as non-volume and non-bucket owner should fail");
```

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestBucketOwner.java [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+    } catch (Exception ex) {
+      LOG.info(ex.getMessage());
+    }
+    //Key Delete - should fail
+    try {
+      OzoneVolume volume = cluster.getClient().getObjectStore()
+              .getVolume("volume1");
+      ozoneBucket = volume.getBucket("bucket1");
+      ozoneBucket.deleteKey("key2");
+      fail();
```
Comment (smengcl) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestBucketOwner.java:139]:
```suggestion
      fail("Delete key as non-volume and non-bucket owner should fail");
```

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestBucketOwner.java [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+    } catch (Exception ex) {
+      LOG.info(ex.getMessage());
+    }
+    //Key Rename - should fail
+    try {
+      OzoneVolume volume = cluster.getClient().getObjectStore()
+              .getVolume("volume1");
+      ozoneBucket = volume.getBucket("bucket1");
+      ozoneBucket.renameKey("key2", "key4");
+      fail();
```
Comment (smengcl) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestBucketOwner.java:149]:
```suggestion
      fail("Rename key as non-volume and non-bucket owner should fail");
```

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmAcls.java [hunk 12]
```diff
@@ -132,7 +132,7 @@ public void testBucketCreationPermissionDenied() throws Exception {
         () -> volume.createBucket(bucketName));
 
     assertTrue(logCapturer.getOutput()
-        .contains("doesn't have CREATE permission to access bucket"));
+        .contains("doesn't have READ permission to access volume"));
```
[the remark this answers is not in the document]
Comment (aswinshakil) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmAcls.java:135]: The test doesn't use `OzoneNativeAuthorizer`. The test has its own authorizer `OzoneAccessAuthorizerTest`. Apart from `OzoneNativeAuthorizer`, other Authorizers which implements `IAccessAuthorizer` the parent level access required is READ, i.e similar to Ranger. Should this be changed?

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java [hunk 13]
```diff
@@ -473,6 +473,8 @@ public void createBucket(
     verifyCountsQuota(bucketArgs.getQuotaInNamespace());
     verifySpaceQuota(bucketArgs.getQuotaInBytes());
 
+    String owner = bucketArgs.getOwner() == null ?
+            ugi.getShortUserName() : bucketArgs.getOwner();
```
[the remark this answers is not in the document]
Comment (aswinshakil) [reply hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java:477]: Yes we fall back to the current ugi as bucket owner. This is similar to how we do it `createVolume` 
https://github.com/apache/ozone/blob/e658dff34fc917bfc3504332a6ed6945b28717de/hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java#L313-L314

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneAclUtils.java [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+      OzoneObj.StoreType storeType, IAccessAuthorizer.ACLType aclType,
+      String vol, String bucket, String key, String volOwner,
+      String bucketOwner, UserGroupInformation user, InetAddress remoteAddress,
+      String hostName) throws IOException {
+
+    boolean isVolOwner = isOwner(user, volOwner);
+
+    IAccessAuthorizer.ACLType parentAclRight = aclType;
+
+    if(ozoneManager.isNativeAuthorizerEnabled()) {
```
[the remark this answers is not in the document]
Comment (aswinshakil) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneAclUtils.java:63]: Thank you for the suggestions I will update the PR.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneAclUtils.java [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+        parentAclRight = IAccessAuthorizer.ACLType.READ;
+      }
+    } else {
+      parentAclRight =  IAccessAuthorizer.ACLType.READ;
+
+    }
+
+    switch (resType) {
+    case VOLUME:
+      ozoneManager.checkAcls(resType, storeType, aclType, vol, bucket, key,
```
[the remark this answers is not in the document]
Comment (aswinshakil) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneAclUtils.java:79]: Sure, will add them.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneAclUtils.java [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+          user, remoteAddress, hostName, true,
+          volOwner);
+      break;
+    case BUCKET:
+    case KEY:
+    case PREFIX:
+      ozoneManager.checkAcls(OzoneObj.ResourceType.VOLUME, storeType,
```
[the remark this answers is not in the document]
Comment (aswinshakil) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneAclUtils.java:86]: Noted. I will update the PR.

**Comment by aswinshakil** [comment]
Thank you for the review. I will update the PR with the suggested changes. For the above questions,

1. Yes, I will resolve the Merge Conflict. 
2. All the `checkAcls` call is already navigated to `OzoneAclUtils.checkAllAcls()`
3. The comments for the previous PR have been addressed.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmAcls.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (smengcl) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmAcls.java:135]: Ah okay, nvm, I thought the test was using `OzoneNativeAuthorizer`. It is fine in this case then.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneAclUtils.java [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+      if (aclType == IAccessAuthorizer.ACLType.CREATE ||
+          aclType == IAccessAuthorizer.ACLType.DELETE ||
+          aclType == IAccessAuthorizer.ACLType.WRITE_ACL) {
+        parentAclRight = IAccessAuthorizer.ACLType.WRITE;
+      } else if (aclType == IAccessAuthorizer.ACLType.READ_ACL ||
+          aclType == IAccessAuthorizer.ACLType.LIST) {
+        parentAclRight = IAccessAuthorizer.ACLType.READ;
+      }
+    } else {
+      parentAclRight =  IAccessAuthorizer.ACLType.READ;
```
Comment (prashantpogde) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneAclUtils.java:76]: Simplify If conditions ?

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java [hunk 18]
```diff
@@ -234,15 +233,66 @@ protected void checkACLs(OzoneManager ozoneManager, String volumeName,
...
+        if (aclType == IAccessAuthorizer.ACLType.CREATE ||
+                aclType == IAccessAuthorizer.ACLType.DELETE ||
+                aclType == IAccessAuthorizer.ACLType.WRITE_ACL) {
+          parentAclRight = IAccessAuthorizer.ACLType.WRITE;
+        } else if (aclType == IAccessAuthorizer.ACLType.READ_ACL ||
+                aclType == IAccessAuthorizer.ACLType.LIST) {
+          parentAclRight = IAccessAuthorizer.ACLType.READ;
+        }
+      } else {
+        parentAclRight = IAccessAuthorizer.ACLType.READ;
```
Comment (prashantpogde) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java:262]: if/else can be simplified.

**Review by prashantpogde — COMMENTED** [review]
Can we also write some test cases for different vol-owner/bucketowner combinations to test different if/else conditions in checkACLs() ?

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java [hunk 19]
```diff
@@ -513,6 +513,8 @@ public void createBucket(
     verifyCountsQuota(bucketArgs.getQuotaInNamespace());
     verifySpaceQuota(bucketArgs.getQuotaInBytes());
 
+    String owner = bucketArgs.getOwner() == null ?
```
[the remark this answers is not in the document]
Comment (aswinshakil) [reply hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java:516]: Yes, we can assign a bucket owner when we create a bucket in `CreateBucketHandler` with --user tag. This is only for bucket creation, we check bucket owner in `OzoneAclUtils.checkAllAcls` for other buckets related access.

**Review by smengcl — APPROVED** [review]

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneAclUtils.java (hunk 17 above)
Comment (prashantpogde) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneAclUtils.java:76]: ??
if(ozoneManager.isNativeAuthorizerEnabled()) {
      if (aclType == IAccessAuthorizer.ACLType.CREATE ||
          aclType == IAccessAuthorizer.ACLType.DELETE ||
          aclType == IAccessAuthorizer.ACLType.WRITE_ACL) {
        parentAclRight = IAccessAuthorizer.ACLType.WRITE;
    } else {
      parentAclRight =  IAccessAuthorizer.ACLType.READ;
    }

**Review by prashantpogde — APPROVED** [review]
LGTM

**Pull request merged**